polygeist/test/polygeist-opt/sycl/convertToLLVMABI.mlir

```diff
@@ -26,3 +28,41 @@ gpu.func @return() -> (memref<?xi32, 1>, i64) {
   gpu.return %memref, %c0: memref<?xi32, 1>, i64
 }
 }
+
+// -----
+
+// CHECK:  func.func @caller([[A0:.*]]: !llvm.ptr<i32, 1>) -> !llvm.ptr<i32, 1> {
+// CHECK-NEXT:    [[P2M:%.*]] = "polygeist.pointer2memref"([[A0]]) : (!llvm.ptr<i32, 1>) -> memref<?xi32, 1>
+// CHECK-NEXT:    [[M2P:%.*]] = "polygeist.memref2pointer"([[P2M]]) : (memref<?xi32, 1>) -> !llvm.ptr<i32, 1>
+// CHECK-NEXT:    [[SYCLCALL:%.*]] = sycl.call([[M2P]]) {Function = @callee, MangledName = @callee} : (!llvm.ptr<i32, 1>) -> !llvm.ptr<i32, 1>
+// CHECK-NEXT:    [[P2M:%.*]] = "polygeist.pointer2memref"([[SYCLCALL]]) : (!llvm.ptr<i32, 1>) -> memref<?xi32, 1>
+// CHECK-NEXT:    [[M2P:%.*]] = "polygeist.memref2pointer"([[P2M]]) : (memref<?xi32, 1>) -> !llvm.ptr<i32, 1>
+// CHECK-NEXT:    [[FUNCCALL:%.*]] = call @callee([[M2P]]) : (!llvm.ptr<i32, 1>) -> !llvm.ptr<i32, 1>
+// CHECK-NEXT:    [[P2M:%.*]] = "polygeist.pointer2memref"([[FUNCCALL]]) : (!llvm.ptr<i32, 1>) -> memref<?xi32, 1>
+// CHECK-NEXT:    [[M2P:%.*]] = "polygeist.memref2pointer"([[P2M]]) : (memref<?xi32, 1>) -> !llvm.ptr<i32, 1>
+// CHECK-NEXT:    return [[M2P]] : !llvm.ptr<i32, 1>
+
+func.func @caller(%arg0: memref<?xi32, 1>) -> memref<?xi32, 1> { 
+  %syclcall = sycl.call(%arg0) {Function = @callee, MangledName = @callee} : (memref<?xi32, 1>) -> memref<?xi32, 1>
+  %funccall = func.call @callee(%syclcall) : (memref<?xi32, 1>) -> memref<?xi32, 1>
+  func.return %funccall: memref<?xi32, 1>
+}
+
+// CHECK: func.func private @callee(!llvm.ptr<i32, 1>) -> !llvm.ptr<i32, 1>
+func.func private @callee(%arg0: memref<?xi32, 1>) -> memref<?xi32, 1>
+
+// -----
+
+// CHECK: !sycl_id_1_ = !sycl.id<1>
+// CHECK: func.func @constructor_caller([[A0:%.*]]: !llvm.ptr<!sycl_id_1_, 1>) {
+// CHECK-NEXT:    [[P2M:%.*]] = "polygeist.pointer2memref"([[A0]]) : (!llvm.ptr<!sycl_id_1_, 1>) -> memref<?x!sycl_id_1_, 1>
+// CHECK-NEXT:    [[M2P:%.*]] = "polygeist.memref2pointer"([[P2M]]) : (memref<?x!sycl_id_1_, 1>) -> !llvm.ptr<!sycl_id_1_, 1>
```

Comment:
TODO: Add canonicalization rules for this.

Reply:
`Memref2Pointer2MemrefCast` and `Memref2PointerIndex` exist in `llvm/polygeist/lib/polygeist/Ops.cpp`. They simplify `pointer2memref(memref2pointer(x))` and `pointer2memref(memref2pointer(x))` respectively. There exists a `CanonicalizerPass` after `ConvertToLLVMABIPass`, so these kinds of patterns should be able to be simplified when running the pipeline.